### PR TITLE
Make statement logging persistent

### PR DESCRIPTION
**Pull request by umanwizard**
This stores the statement log in storage-managed collections, rather than in system tables, so they won't be refreshed and cleared on startup. In order to avoid unnecessarily high load on Cockroach, we buffer the updates and flush them every 5 seconds.

On startup, we re-ingest the log in order to truncate the parts of it that are older than a configurable value, by default 30 days. Currently, this has the potential to use significant memory; we may need to switch to a more incremental approach. We will need to watch this closely when turning on statement logging in prod.

### Motivation

  * This PR adds a known-desirable feature.

    persisted statement logging

### Tips for Reviewer

By far the most complicated part of this is `partially_truncate_statement_logging`. Most of the rest should be straightforward.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Save values in in the statement log for 30 days.

### CODE REVIEW

**Review by philip-stoev — CHANGES_REQUESTED** [review]
Given that this now persists data and new system tables across restats, it needs a Platform Check test to ensure that all works as expected across upgrades. Please let me know if you need any assistance.

File: src/storage-client/src/controller.rs
Comment (pH14) [inline src/storage-client/src/controller.rs:2961]: IMO having an incremental approach here is non-negotiable for this to go to prod. I have https://github.com/MaterializeInc/materialize/pull/20916 up and will polish it up so we can switch to it here

Comment (umanwizard) [reply src/storage-client/src/controller.rs:2961]: When do you think that will be ready? If it will land soon, I'm fine waiting. If it will be a while, I'd rather land this PR now, though; I don't envision it being a problem since we can just avoid turning the `STATEMENT_LOGGING_MAX_SAMPLE_RATE` too high in prod. What do you think?

Comment (pH14) [reply src/storage-client/src/controller.rs:2961]: We can likely get it in this week, so long as your use case here doesn't need to read in consolidated state (I haven't read through the implementation). If we aren't able to get it in, let's add a knob to optionally disable truncation, just so we don't wind up in some scenario where we're crash looping because we're reading everything in

Comment (umanwizard) [reply src/storage-client/src/controller.rs:2961]: >  so long as your use case here doesn't need to read in consolidated state

It does, but I think it can be made to avoid that.

> If we aren't able to get it in, let's add a knob to optionally disable truncation

That works for me

Comment (pH14) [reply src/storage-client/src/controller.rs:2961]: the persist PR has been merged! let us know how it goes

File: src/adapter/src/coord/statement_logging.rs
Comment (ParkMyCar) [inline src/adapter/src/coord/statement_logging.rs:83]: just a thought: if these can be collections end up being very large, it might be nice to use something like [SegVec](https://docs.rs/segvec/latest/segvec/)

File: src/adapter/src/coord/statement_logging.rs
Comment (ParkMyCar) [inline src/adapter/src/coord/statement_logging.rs:114]: This approach totally works, and to me looks good!

An alternative could be to push this logic into the write tasks for storage owned collections, and make them a bit more configurable. i.e. [write_task](https://github.com/ParkMyCar/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs#L174-L178) already waits up to 1 second (`DEFAULT_TICK`) before writing to persist. You could could update write_task so a caller can provide tick and buffer intervals, and then the Coordinator holds onto appender channels. This is definitely an involved change so I'll leave it up to you!

**Review by ParkMyCar — APPROVED** [review]
Things look good from an adapter perspective!

**Comment by umanwizard** [comment]
Hello all, this should be ready for another look if anyone wants to do so before it gets merged.

File: misc/python/materialize/checks/statement_logging.py
Comment (philip-stoev) [inline misc/python/materialize/checks/statement_logging.py:37]: Those sleeps are a bummer -- I recall you had to introduce some in another test as well -- maybe a setting can be implemented that governs the sampling interval , just for the purpose of speeding up the tests?

**Review by philip-stoev — APPROVED** [review]

**Pull request merged**
